src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -1610,11 +1610,24 @@ public void marshal(Object source, HierarchicalStreamWriter w, MarshallingContex
             }
             writeChild(w, context, "iota", e.iota.get(), Integer.class);
             synchronized (e) {
-                for (FlowHead h : e.heads.values()) {
-                    writeChild(w, context, "head", h.getId() + ":" + h.get().getId(), String.class);
+                if (e.headsSerial != null && (e.heads == null || e.heads.isEmpty())) { // Persisting before onLoad has called to load up the real heads
+                    for (Entry<Integer, String> entry : e.headsSerial.entrySet()) {
+                        writeChild(w, context, "head", entry.getKey() + ":" + entry.getValue(), String.class);
+                    }
+                } else {  // onLoad was invoked, we have real heads*/
+                    for (FlowHead h : e.heads.values()) {
+                        writeChild(w, context, "head", h.getId() + ":" + h.get().getId(), String.class);
+                    }
                 }
-                for (BlockStartNode st : e.startNodes) {
-                    writeChild(w, context, "start", st.getId(), String.class);
+
+                if (e.startNodesSerial != null && (e.startNodes == null)) {  // Start nodes have not been lazy-loaded yet
```

Comment:
Is there a reason the clause for `heads` checked `isEmpty` but the clause for `startNodes` does not?

Reply:
Heads should never be empty, start nodes can be if build is completed.

Reply:
(added a comment to explain)